pkg/compact/compact.go

```diff
@@ -309,18 +342,81 @@ func (g *DefaultGrouper) Groups(blocks map[ulid.ULID]*metadata.Meta) (res []*Gro
 			return nil, errors.Wrap(err, "add compaction group")
 		}
 	}
+
+	if g.planner != nil {
+		newGroups := []*Group{}
+		// split each group by available plans
+		for _, group := range res {
+			newGroup, err := g.splitGroupByPlans(group)
+			if err != nil {
+				return nil, err
+			}
+			newGroups = append(newGroups, newGroup...)
+		}
+		res = newGroups
+	}
+
 	sort.Slice(res, func(i, j int) bool {
 		return res[i].Key() < res[j].Key()
 	})
 	return res, nil
 }
 
+// Recursively plan a groups metas until there are no available plans and return a new group for each plan
+// This allows a single group key to utilize group concurrency when there are multiple plans available to compact
+func (g *DefaultGrouper) splitGroupByPlans(group *Group) ([]*Group, error) {
+	groups := []*Group{}
+	// split each group by available plans
+	var seededRand = rand.New(rand.NewSource(time.Now().UnixNano()))
+	for toCompact, err := g.planner.Plan(context.Background(), group.metasByMinTime); len(toCompact) != 0; toCompact, err = g.planner.Plan(context.Background(), group.metasByMinTime) {
```

Comment:
It's hard to read in this long line, but not sure if this works as intended: I would expect here some code that would deduce that the next compaction will output block that can be use in next compaction so we could just do the next one instead 🤔

Reply:
agree hard to read and needs to be refactored. This is to recursively get all available plans for 1 group by removing metas from the parent group on every iteration so eventually the planner has nothing left to plan for that group. It does not deduce future plans as that is a separate issue that needs I wasn't targeting in this feature. Should I?